Add service discovery for IONOS Cloud

xinau opened the pull request:
This change adds a new service discovery for [IONOS Cloud](https://cloud.ionos.com) based on the previous work for Hetzner and Scaleway. The new discovery allows users to create scrape targets based on servers deployed in a IONOS Cloud datacenter.

This is my first PR to this project and only read the contributor guideline, therefore I'm happy to be pointed to any resources I've missed and should have read next.

I've found the implementation (thanks to prior work) straight forward and impact on other things relative small, thus I didn't start a discussion on the mailing list.

Please be aware that this has only been deployed on non-production environments yet.

P.s. Thanks for the awesome work on Prometheus and Ecosystem.

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -256,6 +256,11 @@ hetzner_sd_configs:
 
+
+# List of IONOS service discovery configurations.
+ionos_sd_configs:
```
This should be twice on the page, once for alerting and once for scraping.

xinau replied:
Thanks and fixed.

roidelapluie commented on discovery/ionos/server.go:
```diff
@@ -0,0 +1,170 @@
+		return nil, err
+	}
+
+	// username, password and token are set via http client config
```
```suggestion
	// Username, password and token are set via http client config.
```

xinau replied:
Good catch and fixed.

roidelapluie commented on discovery/ionos/server.go:
```diff
@@ -0,0 +1,170 @@
+		labels := model.LabelSet{
+			serverAvailabilityZoneLabel: model.LabelValue(*server.Properties.AvailabilityZone),
+			serverCPUFamilyLabel:        model.LabelValue(*server.Properties.CpuFamily),
+			serverDatacenterIDLabel:     model.LabelValue(d.datacenterID),
```
Could this come from the API instead? I can see how we could have different values, e.g. if it is case insensitive.

xinau replied:
The API doesn't provide a direct method of returning the datacenter ID from a server or list of servers. But the `servers` object has an `Id` attribute which returns `<datecenter_id>/servers` from which it should be possible to "compute" the datacenter ID. I thought about doing this, but wasn't sure if it would be worth it or not.

roidelapluie replied:
if it is not exposed by the API I think we should expose the ID as is.

roidelapluie replied:
I still think this

xinau replied:
I think I've misunderstood you. Your suggestion is to expose the ID returned by the servers object or not?

roidelapluie replied:
The ID as is, yes

xinau replied:
Done.

roidelapluie commented on discovery/ionos/server.go:
```diff
@@ -0,0 +1,170 @@
+			}
+
+			if len(nicIDs) > 0 {
+				labels[serverPrimaryNICIDLabel] = model.LabelValue(nicIDs[0])
```
It sounds limiting to only have one Ip. can we have the other IP's in other labels?

xinau replied:
Agreed and already implemented. I think you missed the comma seperated labels for ip, nic and lan :). Or am I mistaken?

roidelapluie replied:
In general, this is more useful with dynamic labels `__meta_ionos_server_lan_ip_<lan_id> = <ip address>`

xinau replied:
Thanks for explaining. I've replaced the NIC labels with your suggested dynamic label.

roidelapluie reviewed (commented):
Thanks. Not tested but already done some review.

xinau commented:
@roidelapluie gentle ping. Did you have a chance to test the implementation?

roidelapluie commented:
I am waiting news from IONOS to enable my account

roidelapluie commented on docs/configuration/configuration.md:
````diff
@@ -1533,6 +1538,94 @@ tls_config:
+  from
+
+```yaml
+# The IONOS datacenter ID in which entities should be discovered.
````
```suggestion
# The unique ID of the data center.
```

Can we replace datacenter by data center as that is what is used in IONOS documentation.

roidelapluie commented on documentation/examples/prometheus-ionos.yml:
```diff
@@ -0,0 +1,30 @@
+  - job_name: "node"
+    ionos_sd_configs:
+      - basic_auth:
+          # Replace with IONOS Cloud API Username
```
```suggestion
          # Replace with your username.
```

roidelapluie commented on documentation/examples/prometheus-ionos.yml:
```diff
@@ -0,0 +1,30 @@
+      - basic_auth:
+          # Replace with IONOS Cloud API Username
+          username: username
+          # Replace with IONOS Cloud API Password
```
```suggestion
          # Replace with your password.
```

roidelapluie commented on documentation/examples/prometheus-ionos.yml:
```diff
@@ -0,0 +1,30 @@
+          username: username
+          # Replace with IONOS Cloud API Password
+          password: password
+        # You can find you datacenter ID here: https://dcd.ionos.com/latest/
```
```suggestion
        # You can find you data center unique ID here: https://dcd.ionos.com/latest/
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1533,6 +1538,94 @@ tls_config:
+* `__meta_ionos_server_nic_id`: comma seperated list of the NIC IDs attached to
+  the server
+* `__meta_ionos_server_primary_ip`: the primary IP address of the server
+* `__meta_ionos_server_primary_lan_id`: the primary LAN's ID of the server
```
```suggestion
* `__meta_ionos_server_primary_lan_id`: the ID of the primary LAN
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1533,6 +1538,94 @@ tls_config:
+  the server
+* `__meta_ionos_server_primary_ip`: the primary IP address of the server
+* `__meta_ionos_server_primary_lan_id`: the primary LAN's ID of the server
+* `__meta_ionos_server_primary_nic_id`: the primary NIC's ID of the server
```
```suggestion
* `__meta_ionos_server_primary_nic_id`: the ID of the primary NIC
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1533,6 +1538,94 @@ tls_config:
+
+The labels below depend on the targets boot medium and are mutually exclusive:
+
+* `__meta_ionos_server_boot_cdrom_id`: the CD-ROM's ID the server is booted
```
```suggestion
* `__meta_ionos_server_boot_cdrom_id`: the ID of the CD-ROM the server is booted from
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1533,6 +1538,94 @@ tls_config:
+  from
+* `__meta_ionos_server_boot_image_id`: the ID of the image or snapshot the
+  server is booted from
+* `__meta_ionos_server_boot_volume_id`: the volume's ID the server is booted
```
```suggestion
* `__meta_ionos_server_boot_volume_id`: the ID of the boot volume
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1533,6 +1538,94 @@ tls_config:
+
+* `__meta_ionos_server_boot_cdrom_id`: the CD-ROM's ID the server is booted
+  from
+* `__meta_ionos_server_boot_image_id`: the ID of the image or snapshot the
```
```suggestion
* `__meta_ionos_server_boot_image_id`: the ID of the boot image or snapshot
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1533,6 +1538,94 @@ tls_config:
+* `__meta_ionos_server_state`: the execution state of the server
+* `__meta_ionos_server_type`: the type of the server
+
+The labels below depend on the targets boot medium and are mutually exclusive:
```
```suggestion
```

I would just mix them with the above labels

xinau replied:
Thanks for the suggestions and updated the documentation.

roidelapluie commented on discovery/ionos/server.go:
```diff
@@ -0,0 +1,170 @@
+func (d *serverDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	api := d.client.ServersApi
+
+	servers, _, err := api.DatacentersServersGet(ctx, d.datacenterID).
```
Does this use pagination underneat? Can we ask directly for many servers?

xinau replied:
The [`Servers` struct](https://pkg.go.dev/github.com/ionos-cloud/sdk-go/v6#Servers) includes attribute for [`PaginationLinks`](https://pkg.go.dev/github.com/ionos-cloud/sdk-go/v6@v6.0.5851#PaginationLinks) which can be used to get the next and previous pagination link. 

But, I haven't seen a method provided by the SDK for using the next pagination link. The Servers struct contains an `Offset` and `Limit` attribute, but these only seem to be set when explicitly specified in the API request call. The documentation on the [`DatacentersServersGet` method](https://pkg.go.dev/github.com/ionos-cloud/sdk-go/v6@v6.0.5851#ServersApiService.DatacentersServersGet) says that it returns all servers.

Furthermore, I've taken a look into the official Terraform provider provided by IONOS and they don't seem to perform pagination themselves. https://github.com/ionos-cloud/terraform-provider-ionoscloud/blob/f98403465589e1c3daa7af065db65a4bcaf488e2/ionoscloud/data_source_server.go#L663

So, I'm not sure. My hunch is that it's handled by the SDK or API.

roidelapluie replied:
It's not done by the SDK, but I am fine with doing this as a follow-up PR.

roidelapluie reviewed (commented):
This is a first real round of review. I have access to IONOS now so we can continue working on this. It's not the final review, would be great if you could already address the comments.

roidelapluie commented on discovery/ionos/server.go:
```diff
@@ -118,20 +117,19 @@ func (d *serverDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, er
-			if len(nicIDs) > 0 {
-				labels[serverPrimaryNICIDLabel] = model.LabelValue(nicIDs[0])
-				labels[serverNICIDLabel] = model.LabelValue(join(nicIDs, metaLabelSeparator))
+				name := model.LabelName(serverLANIPLabelPrefix + lan)
```
lan should be the nicID of the lan, mapped with strutil.SanitizeLabelName(uuid)

roidelapluie commented on discovery/ionos/server_test.go:
```diff
@@ -65,12 +65,11 @@ func TestIONOSServerRefresh(t *testing.T) {
 			"__meta_ionos_server_id":                "b501942c-4e08-43e6-8ec1-00e59c64e0e4",
 			"__meta_ionos_server_ip":                ",85.215.233.222,",
 			"__meta_ionos_server_lan_id":            ",1,",
+			"__meta_ionos_server_lan_ip_1":          ",85.215.233.222,",
```
note for this file: `io/ioutil` is in the blacklist: Use corresponding 'os' or 'io' functions instead. (depguard)

xinau replied:
I've replaced the package.

roidelapluie commented on discovery/ionos/server_test.go:
```diff
@@ -61,7 +61,7 @@ func TestIONOSServerRefresh(t *testing.T) {
 			"__meta_ionos_server_boot_cdrom_id":     "0e4d57f9-cd78-11e9-b88c-525400f64d8d",
 			"__meta_ionos_server_cpu_family":        "INTEL_SKYLAKE",
-			"__meta_ionos_server_datacenter_id":     "8feda53f-15f0-447f-badf-ebe32dad2fc0",
+			"__meta_ionos_server_datacenter_id":     "8feda53f-15f0-447f-badf-ebe32dad2fc0/servers",
```
__meta_ionos_server_servers_id

xinau replied:
I've updated the label, thanks for the catch.

roidelapluie commented:
Thanks! I will have a deeper look on monday

xinau commented:
Thanks for a first review and improvements. Looking forward to the next one :)

roidelapluie commented:
The labels still seem confusing to me:

ip addresses should not be in lists.
can one nic belong to multiple lans? it does not seem so.

I would:

Remove _meta_ionos_server_ip (it's equivalent to __address__)
Remove __meta_ionos_server_lan_id
Change "__meta_ionos_server_nic_ip_d09cabed_4fef_425c_8dce_a7f6c8fbe84e" from a list to a single addr.
Remove __meta_ionos_server_primary_ip
Remove __meta_ionos_server_primary_lan_id

If a server does not have a address, it should be dropped

xinau commented:
> can one nic belong to multiple lans? it does not seem so.

No, IONOS allows users to attach multiple NIC's to the same LAN. 

---

> If a server does not have a address, it should be dropped

Indeed, it makes sense to drop servers without a address.


> Remove _meta_ionos_server_ip (it's equivalent to __address__)

The `__address__` label is not the same as `__meta_ionos_server_ip`, but `__meta_ionos_server_primary_ip` + `:port`, where as `__meta_ionos_server_ip` is a list of all IP addresses assigned to a server. Therefore, I would prefer keeping the `__meta_ionos_server_ip` label and remove `__meta_ionos_server_primary_ip`.


What is the reason to change `__meta_ionos_server_nic_ip_<nic_id>` to a single value, I'm asking as IONOS allows users to assign multiple IPs to a single NIC?

xinau commented:
I've seen that it's possible to provide NIC's with a name. 

From a users perspective it makes sense to change `__meta_ionos_server_nic_ip_<id>` to `__meta_ionos_server_nic_ip_<name>. Otherwise, operators which want to set the `__address__` label to the IP address of a specific NIC would have to match UUIDs it's current ID.

The only problem is that the creation of a NIC doesn't require users to set a name for it, therefore leaving it empty by default. This could be resolved by either dropping the dynamic label for NICs with an empty label or defaulting to a NICs ID.

I think it would be fair to drop the dynamic label for NICs without a name, as the creator of the NIC would otherwise have given it a name to be able to make it easier to refer to it later on.

xinau commented:
While adding a test case for a server with multiple NICs, I've noticed that the NICs do not maintain their order between server stop/start. This means that a server with multiple NICs is likely to change it's value for the `__address__` label.

I'm not sure if this is an issue or not, but also haven't found a NIC attribute to sort on. Both `deviceNumber` and `pciSlot` will be reassigned when a NIC is deleted and a new one is created.

Additionaly, when a new NIC is added it will be prepended on the list of NICs.

I'm not sure if this is a problem and if it needs to be solved, but wanted to shared my findings.

roidelapluie commented:
> Additionaly, when a new NIC is added it will be prepended on the list of NICs.

Then we probably should invert the list.

xinau commented:
I've seen, I've misread your question regarding the possibility to attach multiple LANs to a single NIC. :facepalm: 

It's not possible to attach a single NIC to multiple LAN's, but as above multiple NIC's to a single LAN.

xinau commented:
To add to my comment above regarding NIC ordering.

Today, I've noticed that the Data Center Designer provides users the option to sort NICs by either Device number (default) or PCI slot. This ensures that the order of NICs remains the same across instance restarts and when a new NIC is added, as long as previously no NIC has been deleted as the device number would be reassigned.

![image](https://user-images.githubusercontent.com/30144822/167302110-109be6f0-1463-4556-bceb-1d79d19736f6.png)

xinau commented:
With the current changes I've

* added dropping of targets with no IP address
* removed the `__meta_ionos_primary_ip` label
* removed the `__meta_ionos_server_lan_id` and `__meta_ionos_server_primary_lan_id` labels
* changed the dynamic label `__meta_ionos_server_nic_ip_<nic_id>` to `__meta_ionos_server_nic_ip_<nic_name>` by grouping NIC IPs with the same NIC name. NIC's without a name are grouped under `unnamed`.
* inverted the list of NICs before labeling
* added test cases for instances with multiple NICs

roidelapluie commented:
Thanks! I have reviewed this and tested with IONOS cloud, and it looks good to me!

xinau commented:
Thanks for the review and help getting this landed. :)
